### PR TITLE
Add pageTrueLoad method to global

### PR DESCRIPTION
**Pull request by DemianParkhomenko**

### CODE REVIEW

File: src/browser/chrome.ts [hunk 1]
```diff
@@ -40,16 +40,54 @@ export const closeBrowser = async () => {
...
+            '${LAST_RESPONSE_AT_VARIABLE_NAME}'
+            );
+          const counterRequestsUnfinished = +localStorage.getItem(
+            '${COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME}'
+          );
+          const isLoaded = lastResponseAt && Date.now() - lastResponseAt > idleTimeout;
+          if (isLoaded && counterRequestsUnfinished === 0) {
+            resolve();
+            break;
+          } else {}
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:74]: A sync while loop will just block the thread forever, you need to add some kind of a timeout or await promise

File: src/browser/chrome.ts [hunk 2]
```diff
@@ -157,8 +191,40 @@ export const openUrl = async ({
...
+
+    let counterRequestsUnfinished = 0;
+    const updateCounter = async (counter: number) => { 
+      counterRequestsUnfinished = counter;
+      try {
+        await page.evaluate(({ counterRequestsUnfinished, COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME }) => {
+          const item = counterRequestsUnfinished.toString();
+          localStorage.setItem(COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME, item);
+        }, { counterRequestsUnfinished, COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME });
+      } catch (e) {}
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:203]: Silencing errors can make us debug it. Better to `console.error` if you want soft error handling

File: src/browser/chrome.ts [hunk 3]
```diff
@@ -157,8 +191,40 @@ export const openUrl = async ({
...
+    };
+    const onRequestProcessed = async () => {
+      try {
+        await page.evaluate(({ LAST_RESPONSE_AT_VARIABLE_NAME }) => {
+          const item = Date.now().toString();
+          localStorage.setItem(LAST_RESPONSE_AT_VARIABLE_NAME, item);
+        }, {LAST_RESPONSE_AT_VARIABLE_NAME});
+      } catch (e) {}
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:211]: Same

File: src/browser/chrome.ts [hunk 4]
```diff
@@ -157,8 +191,40 @@ export const openUrl = async ({
...
+     };
+
+    page.on('request', async () => {
+      await updateCounter(counterRequestsUnfinished + 1);
+    });
+    page.on('requestfailed', async () => { 
+      await updateCounter(counterRequestsUnfinished - 1);
+      await onRequestProcessed();
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:219]: Instead of 2 functions and evaluates, you could have just one function with a flag, for instance, 

```ts
await updateCounter(counterRequestsUnfinished - 1, {
  requestEnd: true
});
```

or even more smarter,

```ts
await updateCounter(-1);
```

That's all you need. And the function would abstract the rest (inc/dec + setting vars). 

Please simplify

File: src/browser/chrome.ts [hunk 5]
```diff
@@ -157,8 +191,40 @@ export const openUrl = async ({
         msg.type() === "error" &&
         log(`JS error |`, msg.text().replace(/\n[\w\W]*/, ""), `@ [${url}]`)
     );
+
+    let counterRequestsUnfinished = 0;
+    const updateCounter = async (counter: number) => { 
+      counterRequestsUnfinished = counter;
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:197]: Reassign is confusing. Just pass the delta as an argument, `-1` or `1` and inc/dec

File: src/browser/chrome.ts [hunk 6]
```diff
@@ -40,16 +40,49 @@ export const closeBrowser = async () => {
...
+  timeout: number
+) => {
+  const PAGE_TRUE_LOAD = `
+  async pageTrueLoad ({
+    idleTimeout = 2000, resolveBeforeRenderingTimeout = 2000
+  } = {}) {
+    const timeoutPromise = new Promise((resolve) => {
+      setTimeout(() => {
+        resolve();
+      }, ${timeout} - resolveBeforeRenderingTimeout);
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:57]: `timeout` var won't work reliably here. You need something like `timeoutAt` (absolute time) passed from the higher context.

say the api user requested `timeout=30000`

By the time this code is invoked, you [could have already spent](https://github.com/nikitaeverywhere/webrender/blob/9f91782732e20479d0f05bf3c607a67303f546e6/src/browser/chrome.ts#L179) for instance 4 seconds navigating to the page etc. 

Please ensure this timeout is set relatively to the absolute rendering end timestamp.

File: src/browser/chrome.ts [hunk 7]
```diff
@@ -40,16 +40,49 @@ export const closeBrowser = async () => {
...
+  jsOn: JsOn = "commit",
+  timeout: number
+) => {
+  const PAGE_TRUE_LOAD = `
+  async pageTrueLoad ({
+    idleTimeout = 2000, resolveBeforeRenderingTimeout = 2000
+  } = {}) {
+    const timeoutPromise = new Promise((resolve) => {
+      setTimeout(() => {
+        resolve();
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:56]: I suggest to return something like `{ ok: boolean, timeout: boolean }` indicating whether we have the exit because of the timeout or a "clean exit"

File: src/browser/chrome.ts [hunk 8]
```diff
@@ -40,16 +40,49 @@ export const closeBrowser = async () => {
...
+      addEventListener('load', () => {
+        const checkLoaded = setInterval(() => {
+          const lastResponseAt = +localStorage.getItem('${LAST_RESPONSE_AT_VARIABLE_NAME}');
+          const counterRequestsUnfinished = +localStorage.getItem('${COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME}');
+          const isLoaded = lastResponseAt && Date.now() - lastResponseAt > idleTimeout;
+          if (isLoaded && counterRequestsUnfinished === 0) {
+            clearInterval(checkLoaded);
+            resolve();
+          }
+        }, 1000);
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:70]: Too long, let's use 100ms, these checks are cheap

File: src/browser/chrome.ts [hunk 9]
```diff
@@ -40,16 +40,49 @@ export const closeBrowser = async () => {
...
+        resolve();
+      }, ${timeout} - resolveBeforeRenderingTimeout);
+    });
+
+    const loadPromise = new Promise((resolve) => {
+      addEventListener('load', () => {
+        const checkLoaded = setInterval(() => {
+          const lastResponseAt = +localStorage.getItem('${LAST_RESPONSE_AT_VARIABLE_NAME}');
+          const counterRequestsUnfinished = +localStorage.getItem('${COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME}');
+          const isLoaded = lastResponseAt && Date.now() - lastResponseAt > idleTimeout;
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:65]: `isLoaded` name is confusing because it doesn't account for `counterRequestsUnfinished`. I.e. you send 4 requests, one of them finishes and 3 are pending for more than 2 seconds, then isLoaded becomes true.

File: src/browser/chrome.ts [hunk 10]
```diff
@@ -40,16 +40,49 @@ export const closeBrowser = async () => {
...
+          if (isLoaded && counterRequestsUnfinished === 0) {
+            clearInterval(checkLoaded);
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:67]: nit: also clear timeoutPromise

**Review by nikitaeverywhere — CHANGES_REQUESTED** [review]
You can also update readme in this PR

File: src/browser/chrome.ts (hunk 7 above)
Comment (DemianParkhomenko) [reply src/browser/chrome.ts:56]: What does `ok` mean and how is `ok` different from `timeout`? Maybe we should name it `isTimeouted` or `timeouted`?

Comment (nikitaeverywhere) [reply src/browser/chrome.ts:56]: `isTimeouted` is fine too, perhaps `isTimeout` will sound better. `ok` serves the purpose of "generic ok", because maybe later the error might not only be timeout

File: readme.md [hunk 11]
```diff
@@ -36,6 +36,13 @@ provided as a `"js"` parameter.
 + A result of a custom async JavaScript invocation (`"js"` and `"jsOn"` parameters) is returned in the response as `"result"`.
 + A result of the PDF generation (`"takePdfSnapshot"` parameter) is returned in the response as `"pdfSnapshot"`.
 
+> [!NOTE]  
+> There are utility functions available in the `window` object which can be used in the custom JavaScript code: 
+
+- `async pageTrueLoad ({
+    idleTimeout = 2000, resolveBeforeRenderingTimeout = 2000
+  } = {})` - a promise that resolves when the page is fully loaded (including delayed requests).
```
Comment (nikitaeverywhere) [inline readme.md:44]: I'd rephrase `resolves when there are no more pending network requests, but resolveBeforeRenderingTimeout before the global timeout`

File: readme.md [hunk 12]
```diff
@@ -36,6 +36,13 @@ provided as a `"js"` parameter.
 + A result of a custom async JavaScript invocation (`"js"` and `"jsOn"` parameters) is returned in the response as `"result"`.
 + A result of the PDF generation (`"takePdfSnapshot"` parameter) is returned in the response as `"pdfSnapshot"`.
 
+> [!NOTE]  
+> There are utility functions available in the `window` object which can be used in the custom JavaScript code: 
+
+- `async pageTrueLoad ({
```
Comment (nikitaeverywhere) [inline readme.md:42]: webrender.pageTrueLoad

File: readme.md (hunk 11 above)
Comment (nikitaeverywhere) [reply readme.md:44]: BTW maybe a better name is `await webrender.pendingRequests()` for this function
Up to you

File: src/browser/chrome.ts [hunk 13]
```diff
@@ -40,16 +40,49 @@ export const closeBrowser = async () => {
...
+  const WEBRENDER = `const webrender = {${PAGE_TRUE_LOAD}};`;
+
+  return `try {
+   const AsyncFunction = Object.getPrototypeOf(
     async function () {}
   ).constructor;
-  const f = new AsyncFunction("webrender", \`${js
-    .replace(/`/g, "\\`")
-    .replace(/\${/g, "\\${")}\`);
+  const f = new AsyncFunction("_webrender", \`${WEBRENDER + js .replace(/`/g, "\\`")
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:84]: `js .repl` (space)

File: src/browser/chrome.ts [hunk 14]
```diff
@@ -157,8 +191,53 @@ export const openUrl = async ({
...
+            const item = counterRequestsUnfinished.toString();
+            localStorage.setItem(COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME, item);
+          },
+          { counterRequestsUnfinished, COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME }
+        );
+      } catch (e) {
+        console.error(e);
+      }
+      if (inc < 0) {
+        await onRequestProcessed();
```
Comment (nikitaeverywhere) [inline src/browser/chrome.ts:223]: instead of page.evaluate 2 times (which is somewhat heavy and less readable I think), please refactor to use a single page.evaluage 

```ts
await page.evaluate(
          ({ counterRequestsUnfinished, COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME, inc }) => {
            const item = counterRequestsUnfinished.toString();
            localStorage.setItem(COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME, item);
            if (inc < 0) {
              const item2 = Date.now().toString();
              localStorage.setItem(LAST_RESPONSE_AT_VARIABLE_NAME, item2);
            }
          },
          { counterRequestsUnfinished, COUNTER_REQUEST_UNFINISHED_VARIABLE_NAME, inc }
        );
```

**Review by nikitaeverywhere — APPROVED** [review]
Approved given comments fixed

**Review by nikitaeverywhere — APPROVED** [review]
